File: space-game/Core/Extensions/EnumerableExtensions.vb
Imports System.Runtime.CompilerServices
Imports Core.Classes
Imports Core.Enums

Namespace Extensions
    Public Module EnumerableExtensions

        <Extension>
        Public Function HasAny(Of T)(this As IEnumerable(Of T)) As Boolean
            Return this IsNot Nothing AndAlso this.Any()
        End Function

        <Extension>
        Public Function GetHighestAvailableResolution(this As IEnumerable(Of Texture), type As TextureType) As String
            Dim result = ""

            Dim r = GetTexture(this, type).
                    OrderByDescending(Function(t) t.QualityEnum).
                    FirstOrDefault()

            If r IsNot Nothing Then
                result = r.Path
            End If

            Return result
        End Function

        <Extension>
        Public Function GetHighestAvailableResolutionForEachType(this As IEnumerable(Of Texture)) As List(Of Texture)
            Dim result As New Dictionary(Of TextureType, Texture)

            Dim textures = this.
                            OrderByDescending(Function(t) t.QualityEnum).
                            ToList()

            For i = 0 To textures.Count - 1 Step 1
                Dim t = textures(i)
                If Not String.IsNullOrWhiteSpace(t.Path) AndAlso Not result.ContainsKey(t.TypeEnum) Then
                    result.Add(t.TypeEnum, t)
                End If
            Next

            Return result.Values.ToList()
        End Function

        <Extension>
        Public Function GetLowestAvailableResolution(this As IEnumerable(Of Texture), type As TextureType) As String
            Dim result = ""

            Dim r = GetTexture(this, type).
                    OrderBy(Function(t) t.QualityEnum).
                    FirstOrDefault()

            If r IsNot Nothing Then
                result = r.Path
            End If

            Return result
        End Function

        Private Function GetTexture(textures As IEnumerable(Of Texture), type As TextureType) As IEnumerable(Of Texture)
            Return textures.Where(Function(t) t.TypeEnum.Equals(type) AndAlso Not String.IsNullOrWhiteSpace(t.Path))
        End Function

        <Extension>
        Public Function IsEquivalent(this As IEnumerable(Of Texture), other As IEnumerable(Of Texture)) As Boolean
            If (this Is Nothing And other Is Nothing) OrElse (this IsNot Nothing And other IsNot Nothing) Then
                If this.Count.Equals(other.Count) Then

                    Dim sortedThis = this.OrderBy(Function(l) l.Path).
                                                ThenBy(Function(l) l.Quality).
                                                ThenBy(Function(l) l.Type)

                    Dim sortedOther = other.OrderBy(Function(l) l.Path).
                                            ThenBy(Function(l) l.Quality).
                                            ThenBy(Function(l) l.Type)

                    Return sortedThis.SequenceEqual(sortedOther)
                End If
            End If

            Return False
        End Function
    End Module
End Namespace

File: space-game/Core/Extensions/EnumerableExtensions.cs
using Core.Classes;
using Core.Enums;
using System.Collections.Generic;
using System.Linq;

namespace Core.Extensions
{
    public static class EnumerableExtensions
    {
        public static bool HasAny<T>(this IEnumerable<T> @this)
        {
            return @this != null && @this.Any();
        }

        public static string GetHighestAvailableResolution(this IEnumerable<Texture> @this, TextureType type)
        {
            string result = "";

            Texture r = GetTexture(@this, type).OrderByDescending(t => t.QualityEnum).FirstOrDefault();

            if (r != null)
            {
                result = r.Path;
            }

            return result;
        }

        public static List<Texture> GetHighestAvailableResolutionForEachType(this IEnumerable<Texture> @this)
        {
            Dictionary<TextureType, Texture> result = new Dictionary<TextureType, Texture>();

            List<Texture> textures = @this.OrderByDescending(t => t.QualityEnum).ToList();

            for (int i = 0; i <= textures.Count() - 1; i += 1)
            {
                Texture t = textures[i];
                if (!string.IsNullOrWhiteSpace(t.Path) && !result.ContainsKey(t.TypeEnum))
                {
                    result.Add(t.TypeEnum, t);
                }
            }

            return result.Values.ToList();
        }

        public static string GetLowestAvailableResolution(this IEnumerable<Texture> @this, TextureType type)
        {
            string result = "";

            Texture r = GetTexture(@this, type).OrderBy(t => t.QualityEnum).FirstOrDefault();

            if (r != null)
            {
                result = r.Path;
            }

            return result;
        }

        private static IEnumerable<Texture> GetTexture(IEnumerable<Texture> textures, TextureType type)
        {
            return textures.Where(t => t.TypeEnum.Equals(type) && !string.IsNullOrWhiteSpace(t.Path));
        }

        public static bool IsEquivalent(this IEnumerable<Texture> @this, IEnumerable<Texture> other)
        {
            if ((@this == null & other == null) || (@this != null & other != null))
            {

                if (@this.Count().Equals(other.Count()))
                {
                    List<Texture> sortedThis = @this.OrderBy(l => l.Path).ThenBy(l => l.Quality).ThenBy(l => l.Type).ToList();

                    List<Texture> sortedOther = other.OrderBy(l => l.Path).ThenBy(l => l.Quality).ThenBy(l => l.Type).ToList();

                    return sortedThis.SequenceEqual(sortedOther);
                }
            }

            return false;
        }
    }
}
